Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx

```diff
@@ -0,0 +1,154 @@
+#include "CTPWorkflowLumi/RawToLumiConverterSpec.h"
+#include "CTPWorkflow/RawToDigitConverterSpec.h"
+#include "CommonUtils/VerbosityConfig.h"
+
+using namespace o2::ctp::lumi_workflow;
+
+void RawToLumiConverterSpec::init(framework::InitContext& ctx)
+{
+}
+
+void RawToLumiConverterSpec::run(framework::ProcessingContext& ctx)
+{
+  mOutputLumiPoints.clear();
+  const gbtword80_t bcidmask = 0xfff;
+  using InputSpec = o2::framework::InputSpec;
+  using ConcreteDataTypeMatcher = o2::framework::ConcreteDataTypeMatcher;
+  using Lifetime = o2::framework::Lifetime;
+  // mOutputHWErrors.clear();
+  std::vector<InputSpec> filter{InputSpec{"filter", ConcreteDataTypeMatcher{"CTP", "RAWDATA"}, Lifetime::Timeframe}};
+  o2::framework::DPLRawParser parser(ctx.inputs(), filter);
+  // setUpDummyLink
+  auto& inputs = ctx.inputs();
+  // if we see requested data type input with 0xDEADBEEF subspec and 0 payload this means that the "delayed message"
+  // mechanism created it in absence of real data from upstream. Processor should send empty output to not block the workflow
+  {
+    static size_t contDeadBeef = 0; // number of times 0xDEADBEEF was seen continuously
+    std::vector<InputSpec> dummy{InputSpec{"dummy", o2::framework::ConcreteDataMatcher{"CTP", "RAWDATA", 0xDEADBEEF}}};
+    for (const auto& ref : o2::framework::InputRecordWalker(inputs, dummy)) {
+      const auto dh = o2::framework::DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+      auto payloadSize = o2::framework::DataRefUtils::getPayloadSize(ref);
+      if (payloadSize == 0) {
+        auto maxWarn = o2::conf::VerbosityConfig::Instance().maxWarnDeadBeef;
+        if (++contDeadBeef <= maxWarn) {
+          LOGP(alarm, "Found input [{}/{}/{:#x}] TF#{} 1st_orbit:{} Payload {} : assuming no payload for all links in this TF{}",
+               dh->dataOrigin.str, dh->dataDescription.str, dh->subSpecification, dh->tfCounter, dh->firstTForbit, payloadSize,
+               contDeadBeef == maxWarn ? fmt::format(". {} such inputs in row received, stopping reporting", contDeadBeef) : "");
+        }
+        ctx.outputs().snapshot(o2::framework::Output{"CTP", "LUMI", 0, o2::framework::Lifetime::Timeframe}, mOutputLumiPoints);
+        return;
+      }
+    }
+    contDeadBeef = 0; // if good data, reset the counter
+  }
+  //
+  float_t counts = 0;
+  uint32_t payloadCTP;
+  uint32_t orbit0 = 0;
+  bool first = true;
+  gbtword80_t remnant = 0;
+  uint32_t size_gbt = 0;
+  for (auto it = parser.begin(); it != parser.end(); ++it) {
+    auto rdh = it.get_if<o2::header::RAWDataHeader>();
+    auto triggerOrbit = o2::raw::RDHUtils::getTriggerOrbit(rdh);
+    if (first) {
+      orbit0 = triggerOrbit;
+      first = false;
+    }
+    auto feeID = o2::raw::RDHUtils::getFEEID(rdh); // 0 = IR, 1 = TCR
+    auto linkCRU = (feeID & 0xf00) >> 8;
+    if (linkCRU == o2::ctp::GBTLinkIDIntRec) {
+      payloadCTP = o2::ctp::NIntRecPayload;
+    } else {
+      return;
+    }
```

Comment:
You are using data of only 1 link, why don't you request only this link in the https://github.com/AliceO2Group/AliceO2/blob/4b086e5f8c575ffea1ee45f59e57a395d8620a8b/Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx#L140 as 
```
inputs.emplace_back("TF", {"CTP", "RAWDATA", <linkID>}, o2::framework::Lifetime::Optional);
```
then you don't need DPLRawParser and this for loop, instead you can directly read 
```
auto ref = inputs.get("TF",<subspec_of_your_link>);
auto rdh =  o2::framework::DataRefUtils::getHeader<o2::header::RAWDataHeader*>(ref);
...
auto payload = pc.inputs().get<const uint8_t>>(ref);
```
Note that in this case if the real input is absent, you will get instead of DEADBEEF stuff the input with the subspec you requested but with empty payload.

Reply:
HI @shahor02 ,
inputs.emplace_back("TF", {"CTP", "RAWDATA", <linkID>}, o2::framework::Lifetime::Optional);
does not compile - there is no such function.
I changed it to
inputs.emplace_back("TF", "CTP", "RAWDATA", <linkID>, o2::framework::Lifetime::Optional);
Then when I try to run on MC it can not find data:
[O2/latest-ctpdev-o2] ~/testlumi0 $> o2-raw-file-reader-workflow --input-conf CTPraw.cfg | o2-ctp-lumi-workflow -b 
[ERROR] invalid workflow in o2-ctp-lumi-workflow: No matching output found for CTP/RAWDATA/0 as requested by data processor "CTP-RawStreamDecoder". Candidates:
-DPL/ENUM/1301969186
-CTP/RAWDATA
-FLP/DISTSUBTIMEFRAME/0
-CTP/LUMI/0
? 
Cheers, Roman.

Reply:
sorry, there was a typo: `inputs.emplace_back("TF", "CTP", "RAWDATA", 0, o2::framework::Lifetime::Optional);`

Reply:
Yes, I understood you did typo and fixed it.  Please, read also after 'I changed it ...' in my post above.

Reply:
Ah, right, my comment was wrong: since the RAWDATA is sent with wildcard, the DPL will fail to build a topology with input asking for specific subspec... Indeed, one should use DPLRawParser. Sorry for misleading comment.

Reply:
ok, no problem, thanks. 
Is there a way to specify interaction rate in o2-sim ?

Reply:
It is an option of the `o2-sim-digitizer-workflow`: `--interactionRate <rate_in_Hz>`